commons/src/main/java/fr/free/nrw/commons/upload/GPSExtractor.java

```diff
@@ -33,21 +69,67 @@ public String getCoords() {
         }
 
         if (exif.getAttribute(ExifInterface.TAG_GPS_LATITUDE) == null) {
-            Log.d("Image", "Picture has no GPS info");
-            return null;
-        }
-        else {
+            imageCoordsExists = false;
+            Log.d(TAG, "Picture has no GPS info");
+
+            //Check what user's preference is for automatic location detection
+            boolean gpsPrefEnabled = gpsPreferenceEnabled();
+
+            if (gpsPrefEnabled) {
+                //If pref enabled, get current location
+                Location location = locationManager.getLastKnownLocation(provider);
+
+                if (location != null) {
+                    myLocationListener.onLocationChanged(location);
```

Comment:
The only time getLastKnownLocation() should return a location that you haven't already received should be between when you call requestLocationUpdates and when you're waiting for the first callback into LocationListener.onLocationChanged. I suppose it's possible that getLastKnownLocation would have a more accurate result available but you have the callback frequency set to 400 ms so I doubt it would be significant. If you're going to call getLastKnownLocation every time, I think you should use that value directly and don't bother setting currentLatitude / Longitude members in the callback. It seems convoluted if you're manually doing the work of invoking the callback each time.